crates/ink/codegen/src/generator/event_def.rs

```diff
@@ -0,0 +1,224 @@
+            let event_variant_ident = ev.ident();
+            let index = ev.index();
+            quote_spanned!(span=>
+                impl ::ink::reflect::EventVariantInfo<#index> for #event_ident {
+                    const NAME: &'static str = ::core::stringify!(#event_variant_ident);
+                    const SIGNATURE_TOPIC: [u8; 32] = ::ink::primitives::event_signature_topic(
+                        <Self as ::ink::reflect::EventInfo>::PATH,
+                        <Self as ::ink::reflect::EventVariantInfo<#index>>::NAME,
+                    );
+                }
+            )
+        });
+        quote_spanned!(span=>
+            #(
+                #impls
+            )*
+        )
+    }
+
+    fn generate_event_metadata_impl(&self) -> TokenStream2 {
+        let event_metadata = super::metadata::EventMetadata::from(self.event_def);
+        event_metadata.generate_code()
+    }
+
+    /// Generate checks to guard against too many topics in event definitions.
+    fn generate_topics_guard(&self) -> TokenStream2 {
+        let span = self.event_def.span();
+        let event_ident = self.event_def.ident();
+        // todo: [AJ] check if event signature topic should be included here (it is now, wasn't before)
+        let len_topics = self.event_def.max_len_topics();
+        let max_len_topics = quote_spanned!(span=>
+            <<#event_ident as ::ink::env::Topics>::Env
+                as ::ink::env::Environment>::MAX_EVENT_TOPICS
+        );
+        quote_spanned!(span=>
+            impl ::ink::codegen::EventLenTopics for #event_ident {
+                type LenTopics = ::ink::codegen::EventTopics<#len_topics>;
+            }
+
+            const _: () = ::ink::codegen::utils::consume_type::<
+                ::ink::codegen::EventRespectsTopicLimit<
+                    #event_ident,
+                    { #max_len_topics },
+                >
+            >();
+        )
+    }
+
+    fn generate_topics_impl(&self) -> TokenStream2 {
+        let span = self.event_def.span();
+        let event_ident = self.event_def.ident();
+
+        let variant_match_arms = self
+            .event_def
+            .variants()
+            .map(|variant| {
+                let span = variant.span();
+                let variant_ident = variant.ident();
+                let (field_bindings, field_topics): (Vec<_>, Vec<_>) = variant.fields()
+                    .filter(|field| field.is_topic)
+                    .map(|field| {
+                        let field_type = field.ty();
+                        let field_ident = field.ident();
+                        let push_topic =
+                            quote_spanned!(span =>
+                                .push_topic::<::ink::env::topics::PrefixedValue<#field_type>>(
+                                    &::ink::env::topics::PrefixedValue {
+                                        // todo: figure out whether we even need to include a prefix here?
```

Comment:
With https://github.com/paritytech/ink/pull/1533, tests fail because of the same topics. 

<img width="930" alt="image" src="https://user-images.githubusercontent.com/18346821/206928104-676a0beb-5b05-4b5c-b0e5-5136cb3f28eb.png">


The `caller` and `from` accounts are the same. It is clearly a problem with the test because the caller and `from` should be different, but it raises discussion about this part.


In the [ethereum](https://docs.alchemy.com/docs/deep-dive-into-eth_getlogs#deciphering-the-response), the first topic is almost always a signature of the event(except manual `LOG*` opcode). And you highlighted it (here)[TODO: link to comment]. So the first topic is affected by the event's name and used types. But other topics are not prefixed and contain raw/hashed(if the value is more than 32 bytes) values.

I think we need to follow the same pattern. The first topic helps to subscribe to a specific event of contracts. With raw/hashed values, we can do searches across all events, only knowing value without knowing the event structure. It is also very helpful.

Prefixing each topic with the field's name allows us not to mix different events or different fields with the same type. And for that, we need to know the definition of the event to do searches. **But**, we can achieve the same by [filtering](https://docs.alchemy.com/docs/deep-dive-into-eth_getlogs#making-a-request-to-eth_getlogs) supported by ethereum nodes(we can do the same). If we need only specific events, we can use the event's signature. If we need a specific field, we can use the `(null, null, 0x...)` format where we want the third topic.

So comparing those two options, the Ethereum approach looks more flexible and performant(we don't need to hash prefixes and values each time). 

It seems you follow the same thoughts(based on your comment), so I only highlighted conclusions here(if not, let's discuss =) ). 
I also created a discussion here because following the Ethereum approach creates several issues we need to solve:

1. How do we want to calculate a unique signature of the event? It should be documented to be clear for developers of tools. We discussed it in this PR before but didn't finalize it. My approach was to use the `hash({Name of the enum}::{Name of the variant})` or `hash({Name of the enum}::{Name of the variant}::{Number of fields})`. It is a bad idea to use types names because, in Rust, we can create a wrapper that behaves as an original type but has a new definition=( We can use only names, as we do for traits. The number of fields is only an idea that we can discuss or maybe that can give you more ideas=) Also we need to define which hash algorithm we want to use. Because we can calculate the hash during event definition, we are not limited.

1. How do we want to process two same topics? Based on the Ethereum approach, we need to allow them, but on the `contract-pallet` side, we don't allow it(and it is an error that I got in the tests).

<img width="730" alt="image" src="https://user-images.githubusercontent.com/18346821/206930282-660c5f63-5b3d-42fd-923c-58b1da64f04f.png">

3. Right now, we sort topics, but it breaks the filtering strategy because the expected topic can be in an unexpected position.